src/bunit.core/Extensions/WaitForHelpers/RenderedFragmentWaitForHelperExtensions.cs

```
@@ -41,6 +41,31 @@ public static void WaitForState(this IRenderedFragmentBase renderedFragment, Fun
 		}
 	}
 
+	/// <summary>
+	/// Wait until the provided <paramref name="statePredicate"/> action returns true,
+	/// or the <paramref name="timeout"/> is reached (default is one second).
+	///
+	/// The <paramref name="statePredicate"/> is evaluated initially, and then each time
+	/// the <paramref name="renderedFragment"/> renders.
+	/// </summary>
+	/// <param name="renderedFragment">The render fragment or component to attempt to verify state against.</param>
+	/// <param name="statePredicate">The predicate to invoke after each render, which must returns <c>true</c> when the desired state has been reached.</param>
+	/// <param name="timeout">The maximum time to wait for the desired state.</param>
+	/// <exception cref="WaitForFailedException">Thrown if the <paramref name="statePredicate"/> throw an exception during invocation, or if the timeout has been reached. See the inner exception for details.</exception>
+	public static async Task WaitForStateAsync(this IRenderedFragmentBase renderedFragment, Func<bool> statePredicate, TimeSpan? timeout = null)
+	{
+		using var waiter = new WaitForStateHelper(renderedFragment, statePredicate, timeout);
+
+		try
+		{
+			await waiter.WaitTask;
```

The `await`s are explicit here for two reasons:
1. We want to see the `WaitFor` in the stack-trace, which would not happen if we just return the `Task` object
2. We can simplify the logic in `catch` as we don't expect any `AggregateException` anymore.

---

Two questions here:

1. For the assertion failure case, do we want the WaitFor to be part of the stack trace? With the timeout, yeah, I guess it makes sense.
2. I would not expect us to need the try/catch, or is it simply enough to just `await waiter.WaitTask`?

---

> 2. I would not expect us to need the try/catch, or is it simply enough to just `await waiter.WaitTask`?

Hmm good question. My way of thinking was that `WaitForXXX` and `WaitForAsyncXXX` should behave the same (more or less). So the only reason keeping the try/catch is the `ExceptionDispatchInfo.Capture(e).Throw();`.

But to be fair this line doesn't make any sense here anymore anyway. We are not mutating the exception in any way thus rendering that line virtually senseless.

I would lean out of the window and would argue that even for the synchronous version there is not much of a gain here. Sure unpacking an `AggregateException` makes perfect sense, but we could just rethrow the inner exception if we want to. No need for `ExceptionDispatchInfo` object.

---

I removed them everywhere, which makes the code much simpler and cleaner (see last two commits).

I guess I could wrap that in a convenient extension method.

---

I like to remove the aggregate exception if there is only one inner exception. That makes things easier for end users. We could have an explicit catch for that exception and let others just flow up.

---

If we await stuff, there is no `AggregateException` anymore in the first place.

But I have to revert one part anyway :D I did too much